File: C# .Net/JDI UI Framework/JDI/Web/Selenium/Elements/Complex/MultiSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Epam.JDI.Commons;
using Epam.JDI.Core.Interfaces.Base;
using Epam.JDI.Web.Selenium.DriverFactory;
using Epam.JDI.Web.Selenium.Elements.APIInteract;
using OpenQA.Selenium;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Complex
{
    public abstract class MultiSelector<TEnum> : BaseSelector<TEnum>, IMultiSelector<TEnum>
        where TEnum : IConvertible
    {
        private string _separator = ", ";
        
        protected MultiSelector(By optionsNamesLocator = null, By allLabelsLocator = null)
            : base(optionsNamesLocator, allLabelsLocator)
        {
            SelectedNameAction = (m, name) => SelectedElementAction(this, GetWebElement(name));
            SelectedNumAction = (m, index) => SelectedElementAction(this, GetWebElement(index));
            GetValueAction = m => AreSelected().Print();
            SetValueAction = (m, value) => SelectListNamesAction(this, value.Split(_separator)); ;
        }

        public Action<MultiSelector<TEnum>> ClearAction = m =>
        {
            if (!m.HasLocator && m.AllLabels == null)
                throw Exception("Can't clear options. No optionsNamesLocator and allLabelsLocator found");
            if (m.Locator.ToString().Contains("%s"))
                throw Exception(
                    "Can't clear options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '{0}')");
            if (m.AllLabels != null)
            {
                m.ClearElements(m.AllLabels.WebElements);
                return;
            }
            var els = m.WebElements;
            if (els.Count == 1)
                m.Selector.DeselectAll();
            else
                m.ClearElements(els);
        };

        private void ClearElements(IList<IWebElement> els)
        {
            els.Where(el => SelectedNameAction(this, el.Text)).ForEach(el => el.Click());
        }

        protected IWebElement GetWebElement(string name)
        {
            if (!HasLocator && AllLabels == null)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            IList<IWebElement> els = null;
            if (Locator.ToString().Contains("%s"))
                els = new GetElementModule {
                    ByLocator = Locator.FillByTemplate(name),
                    Element = this }
                .WebElements;
            if (AllLabels != null)
                els = GetWebElement(AllLabels.WebElements, name);
            if (els == null)
                els = WebElements;
            if (els.Count == 1)
                els = Selector.Options;
            if (els == null)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            els = GetWebElement(els, name);
            if (els == null || els.Count != 1)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            return els[0];
        }

        private IList<IWebElement> GetWebElement(IList<IWebElement> els, string name)
        {
            return els.Where(el => el.Text.Equals(name)).ToList();
        }

        protected IWebElement GetWebElement(int index)
        {
            if (!HasLocator && AllLabels == null)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            if (Locator.ToString().Contains("%s"))
                throw Exception("Can't get options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '{0}')");
            if (AllLabels != null)
                return GetWebElement(AllLabels.WebElements, index);
            var els = WebAvatar.SearchAll().WebElements;
            return GetWebElement(els.Count == 1
                    ? Selector.Options
                    : els, index);
        }

        private IWebElement GetWebElement(IList<IWebElement> els, int index)
        {
            if (index <= 0)
                throw Exception($"Can't get option with index '{index}'. Index should be 1 or more");
            if (index > els.Count)
                throw Exception($"Can't get option with index '{index}'. Found only {els.Count} options");
            return els[index - 1];
        }


        protected Action<MultiSelector<TEnum>, IList<string>> SelectListNamesAction = 
            (m, names) => names.ForEach(name => m.SelectNameAction(m, name));
        protected Action<MultiSelector<TEnum>, IList<int>> SelectListIndexesAction =
            (m, indexes) => indexes.ForEach(index => m.SelectNumAction(m, index));
        
        
        public IMultiSelector<TEnum> SetValuesSeparator(string separator)
        {
            _separator = separator;
            return this;
        }

        public void Select(params string[] names)
        {
            Actions.Select((m, v) => SelectListNamesAction(this, names), names);
        }

        public void Select(params TEnum[] names)
        {
            Select(names.Select(name => name.ToString()).ToArray());
        }

        public void Select(params int[] indexes)
        {
            Actions.Select((m, i) => SelectListIndexesAction(this, indexes), indexes);
        }

        public void Check(params string[] names)
        {
            Clear();
            Select(names);
        }

        public void Check(params TEnum[] names)
        {
            Clear();
            Select(names);
        }

        public void Check(params int[] indexes)
        {
            Clear();
            Select(indexes);
        }

        public void Uncheck(params string[] names)
        {
            CheckAll();
            Select(names);
        }

        public void Uncheck(params TEnum[] names)
        {
            CheckAll();
            Select(names);
        }

        public void Uncheck(params int[] indexes)
        {
            CheckAll();
            Select(indexes);
        }

        public IList<string> AreSelected()
        {
            return Actions.AreSelected(m => Names, (m, name) => SelectedNameAction(this, name));
        }

        public void WaitSelected(params TEnum[] names)
        {
            WaitSelected(names.Select(name => name.ToString()).ToArray());
        }

        public void WaitSelected(params string[] names)
        {
            Actions.WaitSelected((m, n) => Timer.Wait(() => SelectedNameAction(this, n)), names);
        }

        public IList<string> AreDeselected()
        {
            return Actions.AreDeselected(m => Names, n => Timer.Wait(() => SelectedNameAction(this, n)));
        }

        public void WaitDeselected(params TEnum[] names)
        {
            WaitDeselected(names.Select(name => name.ToString()).ToArray());
        }

        public void WaitDeselected(params string[] names)
        {
            Actions.WaitDeselected((m, n) => Timer.Wait(() =>SelectedNameAction(this, n)), names);
        }

        public void Clear()
        {
            Invoker.DoJAction("Clear Options", m => ClearAction(this));
        }

        public void CheckAll()
        {
            Options.Where(label => !SelectedNameAction(this, label)).ForEach(label => SelectNameAction(this, label));
        }
        
        public void SelectAll() 
        {
            CheckAll();
        }
        public void UncheckAll()
        {
            Clear();
        }

    }
}

File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/complex/MultiSelector.java
package com.epam.jdi.uitests.mobile.appium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.EnumUtils;
import com.epam.commons.LinqUtils;
import com.epam.jdi.uitests.core.interfaces.base.IMultiSelector;
import com.epam.jdi.uitests.mobile.appium.driver.WebDriverByUtils;
import com.epam.jdi.uitests.mobile.appium.elements.apiInteract.GetElementModule;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.List;

import static com.epam.commons.LinqUtils.*;
import static com.epam.commons.PrintUtils.print;
import static com.epam.commons.Timer.waitCondition;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by roman.i on 03.10.2014.
 */

public abstract class MultiSelector<TEnum extends Enum> extends BaseSelector<TEnum> implements IMultiSelector<TEnum> {
    private String separator = ", ";

    public MultiSelector() {
        super();
    }

    public MultiSelector(By optionsNamesLocator) {
        super(optionsNamesLocator);
    }

    public MultiSelector(By optionsNamesLocator, By allLabelsLocator) {
        super(optionsNamesLocator, allLabelsLocator);
    }

    protected void clearAction() {
        if (!haveLocator() && allLabels() == null)
            throw exception("Can't clear options. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            throw exception("Can't clear options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '%s')");
        if (allLabels() != null) {
            clearElements(allLabels().getWebElements());
            return;
        }
        List<WebElement> els = getAvatar().searchAll().getElements();
        if (els.size() == 1)
            getSelector().deselectAll();
        else
            clearElements(els);
    }

    private void clearElements(List<WebElement> els) {
        foreach(where(els, el -> isSelectedAction(el.getText())), WebElement::click);
    }

    protected WebElement getElement(String name) {
        List<WebElement> els = null;
        if (!haveLocator() && allLabels() == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            els = new GetElementModule(WebDriverByUtils.fillByTemplate(getLocator(), name), getAvatar().context, this).getElements();
        if (allLabels() != null)
            els = getElement(allLabels().getWebElements(), name);
        if (els == null)
            els = getAvatar().searchAll().getElements();
        if (els.size() == 1)
            els = getSelector().getOptions();
        if (els == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        else
            els = getElement(els, name);
        if (els == null || els.size() != 1)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        return els.get(0);
    }

    private List<WebElement> getElement(List<WebElement> els, String name) {
        return where(els, el -> el.getText().equals(name));
    }

    protected WebElement getElement(int index) {
        if (!haveLocator() && allLabels() == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            throw exception("Can't get options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '%s')");
        if (allLabels() != null)
            return getElement(allLabels().getWebElements(), index);
        List<WebElement> els = getAvatar().searchAll().getElements();
        return getElement(els.size() == 1
                ? getSelector().getOptions()
                : els, index);
    }

    private WebElement getElement(List<WebElement> els, int index) {
        if (index <= 0)
            throw exception("Can't get option with index '%s'. Index should be 1 or more", index);
        if (index > els.size())
            throw exception("Can't get option with index '%s'. Found only %s options", index, els.size());
        return els.get(index - 1);
    }

    protected boolean isSelectedAction(String name) {
        return isSelectedAction(getElement(name));
    }

    protected boolean isSelectedAction(int index) {
        return isSelectedAction(getElement(index));
    }

    protected void selectListAction(String... names) {
        foreach(names, this::selectAction);
    }

    protected void selectListAction(int... indexes) {
        for (int i : indexes) selectAction(i);
    }

    protected String getValueAction() {
        return print(areSelected());
    }

    @Override
    protected void setValueAction(String value) {
        selectListAction(value.split(", "));
    }

    public IMultiSelector<TEnum> setValuesSeparator(String separator) {
        this.separator = separator;
        return this;
    }

    public final void select(String... names) {
        actions.select(this::selectListAction, names);
    }

    public final void select(TEnum... names) {
        select(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void select(int... indexes) {
        actions.select(this::selectListAction, indexes);
    }

    public final void check(String... names) {
        clear();
        select(names);
    }

    public final void check(TEnum... names) {
        clear();
        select(names);
    }

    public final void check(int... indexes) {
        clear();
        select(indexes);
    }

    public final void uncheck(String... names) {
        checkAll();
        select(names);
    }

    public final void uncheck(TEnum... names) {
        checkAll();
        select(names);
    }

    public final void uncheck(int... indexes) {
        checkAll();
        select(indexes);
    }

    public final List<String> areSelected() {
        return actions.areSelected(this::getNames, this::isSelectedAction);
    }

    public final void waitSelected(TEnum... names) {
        waitSelected(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void waitSelected(String... names) {
        actions.waitSelected(n -> waitCondition(() -> isSelectedAction(n)), names);
    }

    public final List<String> areDeselected() {
        return actions.areDeselected(this::getNames, n -> waitCondition(() -> isSelectedAction(n)));
    }

    public final void waitDeselected(TEnum... names) {
        waitDeselected(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void waitDeselected(String... names) {
        actions.waitDeselected(n -> waitCondition(() -> isSelectedAction(n)), names);
    }

    public void clear() {
        invoker.doJAction("Clear Options", this::clearAction);
    }

    public void checkAll() {
        foreach(where(getOptions(), label -> !isSelectedAction(label)), this::selectAction);
    }

}